pkg/cache/tas_flavor_snapshot.go

```diff
@@ -67,110 +70,92 @@ type TASFlavorSnapshot struct {
-		parentFound := false
-		var parent *domain
-		for levelIdx := lastLevelIdx - 1; levelIdx >= 0 && !parentFound; levelIdx-- {
-			parentValues := s.levelValuesPerDomain[childID][:levelIdx+1]
-			parentID := utiltas.DomainID(parentValues)
-			s.levelValuesPerDomain[parentID] = parentValues
-			parent, parentFound = s.domainsPerLevel[levelIdx][parentID]
-			if !parentFound {
-				parent = &domain{
-					sortName: s.sortName(levelIdx, parentID),
-					id:       parentID,
-				}
-				s.domainsPerLevel[levelIdx][parentID] = parent
-			}
-			childDomain.parentID = parentID
-			parent.childIDs = append(parent.childIDs, childID)
-			childID = parentID
-		}
+	for _, node := range s.leaves {
+		s.nodes[node.id] = node
+		s.domainsPerLevel[len(node.levelValues)-1][node.id] = node
+		s.initializeHelper(node)
 	}
 }
 
-func (s *TASFlavorSnapshot) sortName(levelIdx int, domainID utiltas.TopologyDomainID) string {
-	levelValues := s.levelValuesPerDomain[domainID]
-	if len(levelValues) <= levelIdx {
-		s.log.Error(errCodeAssumptionsViolated, "invalid invocation of sortName",
-			"count", len(levelValues),
-			"levelIdx", levelIdx,
-			"domainID", domainID,
-			"levelValuesPerDomain", s.levelValuesPerDomain,
-			"freeCapacityPerDomain", s.freeCapacityPerLeafDomain)
+// helper
+func (s *TASFlavorSnapshot) initializeHelper(dom *domain) {
+	if len(dom.levelValues) == 1 {
+		s.roots[dom.id] = dom
+		return
+	}
+	parentValues := dom.levelValues[:len(dom.levelValues)-1]
+	parentID := utiltas.DomainID(parentValues)
+	parent, parentFound := s.nodes[parentID]
+	if !parentFound {
+		// create parent
+		parent = &domain{
+			id:          parentID,
+			levelValues: parentValues,
+		}
+		s.domainsPerLevel[len(parentValues)-1][parentID] = parent
+		s.nodes[parentID] = parent
 	}
-	// we prefix with the node label value to make it ordered naturally, but
-	// append also domain ID as the node label value may not be globally unique.
-	return s.levelValuesPerDomain[domainID][levelIdx] + " " + string(domainID)
+	// connect parent and child
+	dom.parent = parent
+	parent.children = append(parent.children, dom)
+	s.initializeHelper(parent)
 }
 
 func (s *TASFlavorSnapshot) addCapacity(domainID utiltas.TopologyDomainID, capacity resources.Requests) {
-	s.initializeFreeCapacityPerDomain(domainID)
-	s.freeCapacityPerLeafDomain[domainID].Add(capacity)
+	if s.leaves[domainID].freeCapacity == nil {
+		s.leaves[domainID].freeCapacity = resources.Requests{}
+	}
+	s.leaves[domainID].freeCapacity.Add(capacity)
 }
 
 func (s *TASFlavorSnapshot) addUsage(domainID utiltas.TopologyDomainID, usage resources.Requests) {
-	s.initializeFreeCapacityPerDomain(domainID)
-	s.freeCapacityPerLeafDomain[domainID].Sub(usage)
-}
-
-func (s *TASFlavorSnapshot) initializeFreeCapacityPerDomain(domainID utiltas.TopologyDomainID) {
-	if _, found := s.freeCapacityPerLeafDomain[domainID]; !found {
-		s.freeCapacityPerLeafDomain[domainID] = resources.Requests{}
-	}
+	s.leaves[domainID].freeCapacity.Sub(usage)
```

Comment:
This access was protected by `initializeFreeCapacityPerDomain` , can you provide some convincing argument why this is not needed? Otherwise we cool keep in this PR, because it is sort of orthogonal issue.

Reply:
Reverted